### PR TITLE
feat(log): Introduce Trace log level

### PR DESCRIPTION
**Pull request by svedatbegar**
The logger implementation/interface within Flamingo does not support Trace level logging. This PR adds that functionality.

### CODE REVIEW

**Comment by tessig** [comment]
Hi @svedatbegar 
 
Thanks for implementing this feature. We appreciate your effort and the proposed changes.

Please also adjust the `CueConfig` functions of `zap.Module` and silentzap.Module` to accept `Trace` as log level.

However, the change to the interface `flamingo.Logger` is a breaking change and, strictly seen, this would require us to increase the major version to v4, which is not something we are planning to do in the near future.

I guess it is very unlikely that projects have their own logger implementations in place, but we must still consider it since we strive to maintain backward compatibility as much as possible.

I'd like to pull this topic into a bigger round. We have a maintainers meeting this Thursday and will discuss it there. I will be happy to forward you the invitation if you are interested in taking part, too.

**Comment by svedatbegar** [comment]
Hi @tessig

Thanks for taking a look at the PR. I've made the changes you requested.

If it's necessary to bump the major version for the logger, I'm assuming it could take a while to introduce this feature as you mentioned. But I must add that since trace level logging does not exist in flamingo right now, we'll have to implement or bring in our own logger that has this functionality. We'll have to discuss with the team on how to proceed.

I'd be glad to join the round to discuss this issue.

**Comment by svedatbegar** [comment]
Hello again @tessig ,

I've made the updates you requested. I kept the trace implementations in place while locking cue config updates and the logger interface behind build tags as you suggested. As far as I understand, there are no problems with adding more functionality to the implementations as long as we don't touch the interface.

1 more thing to keep an eye on would be the testing of the trace level logs, as the tests must be copied for them or maybe should be handled differently, but I'm open to suggestions.

**Comment by tessig** [comment]
Very nice work @svedatbegar 🙇 

I agree, the example tests are a problem, because godoc does not support build tags so far: https://github.com/golang/go/issues/3398

With zap.Logger's Trace function being tested, I would be fine with having the Trace not in the ExampleTest for now, or a duplicated ExampleTest (which will not render in godoc for now).

Also, I feel the need that we should mention the tracelog feature and build tag in the documentation. Perhaps we can extend the logging section in `docs/1. Flamingo Basics/5. Flamingo Production.md`. But that can also be done in a separate PR.

File: core/silentzap/levels_standard.go
Comment (carstendietrich) [inline core/silentzap/levels_standard.go:7]: Is there a reason we went for a function? Would prefer a const instead..

**Comment by tessig** [comment]
We do not duplicate the example tests. It's not worth the effort.
Let's just add a commented line, eg
```
// logger.Trace("my", "log", "args") // this is possible with build tag `tracelog`
```

**Review by carstendietrich — CHANGES_REQUESTED** [review]
overall the changes look good to me, would only move the allowed levels to a const instead of a function. After this adjustment, we could merge the changes.

File: core/silentzap/levels_standard.go
Comment (svedatbegar) [reply core/silentzap/levels_standard.go:7]: Perfectly reasonable request. Making the change.

**Review by carstendietrich — APPROVED** [review]

**Review by tessig — APPROVED** [review]

**Comment by tessig** [comment]
Thanks for your contribution, it will be included in v3.9.0. We tidy up documentation and then tag.

**Pull request merged**
